### PR TITLE
Adds ONNX Surrogate support from OMLT

### PR DESCRIPTION
**Pull request by avdudchenko**
## Fixes
This adds ONNX surrogate model object and slightly refactors the current Keras surrogate model to support multiple omlt model types. 

## Summary/Motivation:
Currently, if one generates a NN model using PyTorch, or any other method for use in IDAES and depended project (e.g. WaterTAP) we must export the model to ONNX format, and then to keras format to use in IDAES, which makes little sense. 

This PR adds native support to onnx model format provided through OMLT.  

## Changes proposed in this PR:
-move out part of Keras functions into omlt_surrogate_base.py to enable simple support for multiple omlt import types
-adds ONNXSurrogate object that can load onnx model directly in IDAES flowsheet block 

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by ksbeattie** [comment]
@avdudchenko can you look into the pylint and rtd test failures?

File: idaes/core/surrogate/omlt_base_surrogate_class.py
Comment (rundxdi) [inline idaes/core/surrogate/omlt_base_surrogate_class.py:1]: Changes to this file all look good

File: idaes/core/surrogate/keras_surrogate.py
Comment (rundxdi) [inline idaes/core/surrogate/keras_surrogate.py:1]: Changes to this file all look good

File: idaes/core/surrogate/onnx_surrogate.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+        omlt_scaling, scaled_input_bounds = self.generate_omlt_scaling_objecets()
+
+        # omlt takes *scaled* input bounds as a dictionary with int keys
+        input_bounds = dict(enumerate(self.input_bounds().values()))
+        scaled_input_bounds = omlt_scaling.get_scaled_input_expressions(input_bounds)
+        scaled_input_bounds = {i: tuple(bnd) for i, bnd in scaled_input_bounds.items()}
+
+        # TODO: remove this once new OMLT 1.2 is made available and includes tanh support
+        # overrides default available activation functions for ONNX, tanh is not listed in 1.1 but is supported
+        omltio.onnx_parser._ACTIVATION_OP_TYPES = [
```
Comment (rundxdi) [inline idaes/core/surrogate/onnx_surrogate.py:123]: Does this need catches for anything?

File: idaes/core/surrogate/onnx_surrogate.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+            formulation_object = ReluComplementarityFormulation(net)
+        else:
+            raise ValueError(
+                'An unrecognized formulation "{}" was passed to '
+                "KerasSurrogate.populate_block. Please pass a valid "
+                "formulation.".format(formulation)
+            )
+        self.populate_block_with_net(block, formulation_object)
+
+    def evaluate_surrogate(self, inputs):
```
Comment (rundxdi) [inline idaes/core/surrogate/onnx_surrogate.py:152]: Why is this present but only raises NotImplementError?

**Review by rundxdi — COMMENTED** [review]
@avdudchenko   Generally this all looks good.  I have a couple very minor questions and test failures need to be resolved, then I can approve it.

File: idaes/core/surrogate/onnx_surrogate.py (hunk 2 above)
Comment (avdudchenko) [reply idaes/core/surrogate/onnx_surrogate.py:152]: Left it here for future implementation, but setting up ONNX model to solve outside idaes did not seem trivial at time of first commit. (e.g. there did not seem to be something as simple as _keras_model.predict.

File: idaes/core/surrogate/onnx_surrogate.py (hunk 1 above)
Comment (avdudchenko) [reply idaes/core/surrogate/onnx_surrogate.py:123]: Not sure best catches to implement here, thoughts on just checking for OMLT version? (It should not be needed with future OMLT releases.

**Comment by ksbeattie** [comment]
@avdudchenko @rundxdi should this be part of the next (now May) release?

**Comment by ksbeattie** [comment]
@avdudchenko @rundxdi  now would be a good time to get this on the Aug release, if you can find a reviewer for it

**Comment by ksbeattie** [comment]
@avdudchenko, this is now on the Nov 2024 release board.  Can progress be picked up on this now?

**Comment by ksbeattie** [comment]
@bpaul4, @rundxdi can you review or finish your review?

File: idaes/core/surrogate/onnx_surrogate.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+        This constructor should only be used when first creating the surrogate within IDAES.
+        Once created, this object can be stored to disk using save_to_folder and loaded
+        with load_from_folder
+
+        Args:
+           onnx_model: Onnx model file to be loaded.
+           input_labels: list of str
+              The ordered list of labels corresponding to the inputs in the keras model
+           output_labels: list of str
+              The ordered list of labels corresponding to the outputs in the keras model
```
Comment (bpaul4) [inline idaes/core/surrogate/onnx_surrogate.py:74]: Should this be "in the onnx model"?

File: idaes/core/surrogate/onnx_surrogate.py [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    def populate_block(self, block, additional_options=None):
+        """
+        Method to populate a Pyomo Block with the onnx model constraints.
+
+        Args:
+           block: Pyomo Block component
+              The block to be populated with variables and/or constraints.
+           additional_options: dict or None
+              If not None, then should be a dict with the following keys;
+              'formulation': KerasSurrogate.Formulation
```
Comment (bpaul4) [inline idaes/core/surrogate/onnx_surrogate.py:108]: Should this be "ONNXSurrogate"?

File: idaes/core/surrogate/onnx_surrogate.py [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+        elif formulation == ONNXSurrogate.Formulation.REDUCED_SPACE:
+            formulation_object = ReducedSpaceSmoothNNFormulation(net)
+        elif formulation == ONNXSurrogate.Formulation.RELU_BIGM:
+            formulation_object = ReluBigMFormulation(net)
+        elif formulation == ONNXSurrogate.Formulation.RELU_COMPLEMENTARITY:
+            formulation_object = ReluComplementarityFormulation(net)
+        else:
+            raise ValueError(
+                'An unrecognized formulation "{}" was passed to '
+                "KerasSurrogate.populate_block. Please pass a valid "
```
Comment (bpaul4) [inline idaes/core/surrogate/onnx_surrogate.py:149]: Should this be "ONNXSurrogate"?

File: idaes/core/surrogate/tests/test_onnx_surrogate.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+# onnx, onnx_available = attempt_import("onnx")
+rtol = 1e-4
+atol = 1e-4
+
+
+def load_onnx_model_data(
+    name="net_st_net_5000_STM_100_s_2000000_60_5_tanh_1e-06_4096_tr_15481_Calcite_ST",
+):
+    #    keras_folder_name = os.path.join(this_file_dir(), 'data', 'keras_models', name)
+    #    keras_model = keras.models.load_model(keras_folder_name)
```
Comment (bpaul4) [inline idaes/core/surrogate/tests/test_onnx_surrogate.py:47]: Commented lines, remove if not needed.

**Review by bpaul4 — APPROVED** [review]
I think the changes implemented here are good. Not critical to fix, there are a few places in the ONNX file that refer to "keras" instead of "onnx".

**Comment by dangunter** [comment]
@avdudchenko Can you look at the test failures and see if they can be fixed?

File: idaes/core/surrogate/onnx_surrogate.py (hunk 3 above)
Comment (avdudchenko) [reply idaes/core/surrogate/onnx_surrogate.py:74]: Fixed through out

File: idaes/core/surrogate/onnx_surrogate.py (hunk 4 above)
Comment (avdudchenko) [reply idaes/core/surrogate/onnx_surrogate.py:108]: Fixed

File: idaes/core/surrogate/onnx_surrogate.py (hunk 5 above)
Comment (avdudchenko) [reply idaes/core/surrogate/onnx_surrogate.py:149]: Fixed

File: idaes/core/surrogate/tests/test_onnx_surrogate.py (hunk 6 above)
Comment (avdudchenko) [reply idaes/core/surrogate/tests/test_onnx_surrogate.py:47]: Done

**Pull request merged**